File: src/main/scala/cn/xpleaf/bigdata/spark/scala/core/p2/_01SparkParallelizeRDDOps.scala
package cn.xpleaf.bigdata.spark.scala.core.p2

import org.apache.log4j.{Level, Logger}
import org.apache.spark.rdd.RDD
import org.apache.spark.{SparkConf, SparkContext}

/**
  * 基于并行化集合创建SparkRDD
  */
object _01SparkParallelizeRDDOps {
    def main(args: Array[String]): Unit = {
        val conf:SparkConf = new SparkConf().setMaster("local[2]").setAppName(_01SparkParallelizeRDDOps.getClass.getSimpleName)

        // 屏蔽日志的其它方式，精确屏蔽某一个包产生的日志
        Logger.getLogger("org.apache.spark").setLevel(Level.OFF)
        Logger.getLogger("org.apache.hadoop").setLevel(Level.OFF)

        val sc:SparkContext = new SparkContext(conf)
        val list = List("hello you", "hello he", "hello me")

        /*
        val listRDD:RDD[String] = sc.parallelize(list)
        val wordsRDD:RDD[String] = listRDD.flatMap(line => line.split(" "))
        val pairRDD:RDD[(String, Int)] = wordsRDD.map(word => new Tuple2[String, Int](word, 1))
        val retRDD:RDD[(String, Int)] = pairRDD.reduceByKey((v1, v2) => v1 + v2)
        retRDD.foreach(tuple => println(tuple._1 + "..." + tuple._2))
        */

        // 下面依然是一行代码搞定
        sc.parallelize(list).flatMap(_.split(" ")).map((_, 1)).reduceByKey(_ + _).foreach(t => println(t._1 + "..." + t._2))

        sc.stop()
    }
}

File: src/main/java/cn/xpleaf/bigdata/spark/java/core/p2/_01SparkParallelizeRDDOps.java
package cn.xpleaf.bigdata.spark.java.core.p2;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import scala.Tuple2;

import java.util.Arrays;
import java.util.List;

/**
 * 基于并行化集合创建SparkRDD
 */
public class _01SparkParallelizeRDDOps {
    public static void main(String[] args) {
        SparkConf conf = new SparkConf().setMaster("local[2]").setAppName(_01SparkParallelizeRDDOps.class.getSimpleName());
        JavaSparkContext jsc = new JavaSparkContext(conf);
        jsc.setLogLevel("WARN");    // 设置日志级别，减少控制台输出
        List<String> list = Arrays.asList("hello you", "hello he", "hello me");

        JavaRDD<String> listRDD = jsc.parallelize(list);
        JavaRDD<String> wordsRDD = listRDD.flatMap(line -> {
            return Arrays.asList(line.split(" "));
        });
        JavaPairRDD<String, Integer> pairRDD = wordsRDD.mapToPair(word -> {
            return new Tuple2<String, Integer>(word, 1);
        });
        JavaPairRDD<String, Integer> retRDD = pairRDD.reduceByKey((v1, v2) -> {
            return v1 + v2;
        });
        retRDD.foreach(tuple -> {
            System.out.println(tuple._1 + "..." + tuple._2);
        });

        jsc.close();
    }
}
